src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -27,90 +29,102 @@ protected SwitchBinder(Binder next, SwitchStatementSyntax switchSyntax)
 
         internal static SwitchBinder Create(Binder next, SwitchStatementSyntax switchSyntax)
         {
-            // When pattern-matching is enabled, we use a completely different binder and binding
-            // strategy for switch statements. Once we have confirmed that it is totally upward
-            // compatible with the existing syntax and semantics, we will remove *this* binder
-            // and use the new one for binding all switch statements.
             var parseOptions = switchSyntax?.SyntaxTree?.Options as CSharpParseOptions;
             return
-                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false &&
-                (parseOptions?.Features.ContainsKey("typeswitch") != false || IsPatternSwitch(switchSyntax)))
+                // in C# 6 and earlier, we use the old binder. In C# 7 and later, we use the new binder which
+                // is capable of binding both the old and new syntax. However, the new binder does not yet
+                // lead to a translation that fully supports edit-and-continue.
+                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false || parseOptions?.Features.ContainsKey("typeswitch") != false)
```

Comment:
Users *can* produce such code, by adding, for example "case 0 when false:" to an existing switch. However, rather than duplicating hundreds of existing switch tests with this slight variation and hand-checking their expected behavior, we get the test coverage by adding a smaller set of tests that use that technique, and also running all of the existing tests in this mode (in addition to the normal C# 6 and C# 7 modes) to also exercise these code paths and ensure that we get the expected results. Once we get edit-and-continue working for the new binder, we may be using the new strategy always and we will no longer need this switch for test coverage.

---
In reply to: [70340728](https://github.com/dotnet/roslyn/pull/12416#discussion_r70340728) [](ancestors = 70340728)